presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitSource.java

```diff
@@ -413,6 +450,91 @@ void fail(Throwable e)
         return toCompletableFuture(transform);
     }
 
+    private List<ConnectorSplit> dynamicallyFilterSplits(List<ConnectorSplit> splits)
+    {
+        TupleDomain<HiveColumnHandle> runtimeTupleDomain = getRuntimeTupleDomains();
+
+        Optional<Map<HiveColumnHandle, Domain>> domains = runtimeTupleDomain.getDomains();
+        if (runtimeTupleDomain.isAll() || !domains.isPresent() || domains.get().size() == 0) {
+            return splits;
+        }
+
+        DomainsCache domainsCache = new DomainsCache(domains.get());
+
+        ImmutableList.Builder<ConnectorSplit> result = ImmutableList.builder();
+        for (ConnectorSplit split : splits) {
+            HiveSplit hiveSplit = (HiveSplit) split;
+            List<HivePartitionKey> partitionKeys = hiveSplit.getPartitionKeys();
+            boolean removeSplit = false;
+            for (HivePartitionKey partitionKey : partitionKeys) {
+                Collection<Domain> relevantDomains = domainsCache.getDomains(partitionKey.getName());
+                if (relevantDomains != null && !relevantDomains.isEmpty()) {
+                    Type type = partitionKey.getHiveType().getType(typeManager);
+                    Domain partitionValue;
+                    try {
+                        Object objectToWrite = HiveUtil.parsePartitionValue(partitionKey.getName(), partitionKey.getValue(), type, timeZone).getValue();
+                        partitionValue = Domain.singleValue(type, objectToWrite);
+                    }
+                    catch (PrestoException e) {
+                        // if the type is not supported, skip pruning for that partition
+                        if (e.getErrorCode().equals(NOT_SUPPORTED.toErrorCode())) {
+                            log.warn("Type " + type + " is not supported for Dynamic Partition Pruning");
+                            continue;
+                        }
+
+                        throw e;
+                    }
+                    Predicate<Domain> checkOverlap = d -> d.overlaps(partitionValue);
+                    if (relevantDomains.stream().noneMatch(checkOverlap)) {
+                        log.info("Partition got pruned due to dynamic filters: "
+                                + hiveSplit.getPath() + " for table " + hiveSplit.getDatabase() + "." + hiveSplit.getTable());
+                        removeSplit = true;
+                    }
+                }
+            }
+            if (!removeSplit) {
+                result.add(split);
+            }
+        }
+        return result.build();
+    }
+
+    class DomainsCache
```

Comment:
nit: can this class be defined as `static`?
```suggestion
    static class DomainsCache
```